x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/ssl/SSLReloadDuringStartupIntegTests.java

```diff
@@ -59,66 +81,121 @@ protected boolean transportSSLEnabled() {
+        }
+        try (Timed t = new Timed(LOGGER, Level.INFO, "Ensure fully connected cluster took {}ms.")) {
+            ensureFullyConnectedCluster();
+        }
+    }
+
+    private void waitUntilNodeStartupIsReadyToBegin(final CountDownLatch beforeKeystoreFix) {
+        beforeKeystoreFix.countDown(); // SYNC: Cert update & ES restart
+        try {
+            final long sleepMillis = randomLongBetween(1L, 2000L); // intended sleepMillis
+            try (Timed t = new Timed(LOGGER, Level.INFO, "Awaited {}ms. Sleeping " + sleepMillis + "ms before fixing...")) {
+                beforeKeystoreFix.await(); // SYNC: Cert update & ES restart
+            }
+            try (Timed t = new Timed(LOGGER, Level.INFO, "Slept {}ms, intended " + sleepMillis + "ms. Fixing can start now...")) {
+                Thread.sleep(sleepMillis); // Simulate cert update delay relative to ES start
+            }
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        }
+    }
+
+    private void waitUntilFixKeystoreIsReadyToBegin(final CountDownLatch beforeKeystoreFix) {
+        beforeKeystoreFix.countDown(); // SYNC: Cert update & ES restart
+        try (Timed t = new Timed(LOGGER, Level.INFO, "Awaited {}ms. Node can start now...")) {
+            beforeKeystoreFix.await(); // SYNC: Cert update & ES restart
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        }
     }
 
-    private void copyAndAtomicMoveIfPossible(Path source, Path target) throws IOException {
+    /**
+     * Copy a source file to a destination file, overwriting if necessary.
+     * Use an intermediate temporary file, to attempt an atomic move.
+     * If atomic move fails, fall back to non-atomic move.
+     * @param source File to be copied.
+     * @param target File to be created or overwritten.
+     * @throws IOException Cannot create temp file, or copy source file to temp file, or non-atomic move temp file to target file.
+     */
+    private void atomicCopyIfPossible(Path source, Path target) throws IOException {
+        LOGGER.info("Copying [{}] to [{}]", source, target);
         Path tmp = createTempFile();
+        // LOGGER.info("Temporary file [{}] created", tmp);
         Files.copy(source, tmp, StandardCopyOption.REPLACE_EXISTING);
+        // LOGGER.info("Copy succeeded from [{}] to [{}]", source, tmp);
         try {
             Files.move(tmp, target, StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE);
+            LOGGER.info("Atomic move succeeded from [{}] to [{}]", tmp, target);
         } catch (AtomicMoveNotSupportedException e) {
+            LOGGER.warn("Atomic move failed from [{}] to [{}]", tmp, target, e);
             Files.move(tmp, target, StandardCopyOption.REPLACE_EXISTING);
+            LOGGER.info("Non-atomic move succeeded from [{}] to [{}]", tmp, target);
+        }
+    }
+
+    public static class Timed implements Closeable {
```

Comment:
you can save a bit of boilerplate by making this a record instead of a Class.

Reply:
I think a record requires all class members to be passed in, so this prevents conversion to a record.
https://github.com/elastic/elasticsearch/blob/379a8da3431c85a450c6890b60c28c774ef9cfe9/x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/ssl/SSLReloadDuringStartupIntegTests.java#L185